types.go

```diff
@@ -41,6 +43,128 @@ func (o Strings) MarshalJSON() ([]byte, error) {
 var _ json.Unmarshaler = (*Strings)(nil)
 var _ json.Marshaler = (*Strings)(nil)
 
+// Flag represents a ternary value: false (-1), default (0), or true (+1).
+//
+// When encoded in json, False is "false", Default is "null" (or empty), and True
+// is "true".
+type Flag int8
+
+const (
+	False   Flag = -1
+	Default Flag = 0
+	True    Flag = 1
+)
+
+func (f Flag) MarshalJSON() ([]byte, error) {
+	switch f {
+	case Default:
+		return json.Marshal(nil)
+	case True:
+		return json.Marshal(true)
+	case False:
+		return json.Marshal(false)
+	default:
+		return nil, fmt.Errorf("invalid flag value: %d", f)
+	}
+}
+
+func (f *Flag) UnmarshalJSON(input []byte) error {
+	switch string(input) {
+	case "null", "undefined":
```

Comment:
why both? is this for compatibility with JSON parsers that treat them both as valid equivalents to "empty"?

Reply:
any reason we should also add "default" here in case there's some confusion with the `String()` method below. Or is this a good way to catch people using `String` incorrectly?

Reply:
similarly, what about adding `""`? I guess we don't need to because we have `omitempty` and so we're free to declare `""` as invalid, just wanted to make sure though.

Reply:
> why both? is this for compatibility with JSON parsers that treat them both as valid equivalents to "empty"?

Hm. You're right, undefined is unnecessary.

> any reason we should also add "default" here in case there's some confusion

This doesn't allow strings, only `true`, `false`, and `null`.